File: android/app/src/main/java/com/babelwords/app/ads/AdMobManager.kt
package com.babelwords.app.ads

import android.app.Activity
import android.content.Context
import android.media.AudioManager
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.util.Log
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdRequest
import com.babelwords.app.BabelWordsApplication
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

/**
 * Unified interstitial ad manager (v53 — single pipeline with AtomicBoolean guard,
 * pendingShow auto-show, pre-resume throttle, network callback, session-token safety).
 *
 * All interstitial load/show state lives here — no dual-pipeline duplication.
 *
 * Key methods:
 *   loadInterstitialAndShow()  — primary trigger (loads + shows)
 *   loadRewardedAndShow()      — backward-compat alias → loadInterstitialAndShow()
 *   preloadInterstitial()      — warm cache for faster show
 *   showInterstitial()         — show cached ad (or load+show with pendingShow)
 *   destroy()                  — set isDestroyed, cancel all pending work
 *
 * Events sent to the web app (window.onAdBridgeEvent):
 *   interstitialLoaded / interstitialShown / interstitialClosed / interstitialFailed
 */
class AdMobManager(
    private val context: Context,
    private val eventCallback: (eventType: String, data: String?) -> Unit,
    private val getConsentManager: () -> ConsentManager? = { null },
) {
    private val TAG = "AdMobManager"

    companion object {
        // Sample unit — used ONLY in Firebase Test Lab.
        private const val TEST_INTERSTITIAL_ID = "ca-app-pub-3940256099942544/1033173712"
        private const val MAX_AUTO_SHOW_ATTEMPTS = 3
        // Foreground interstitial cooldown (5 minutes)
        private const val FOREGROUND_AD_COOLDOWN_MS = 5 * 60 * 1000L
        // Ad freshness: 45 min expiry, refresh if older than 40 min
        private const val AD_EXPIRY_MS = 45 * 60 * 1000L
        private const val AD_REFRESH_THRESHOLD_MS = 40 * 60 * 1000L
        // Load throttle: 30s minimum between loads
        private const val LOAD_THROTTLE_MS = 30_000L
        // Load timeout: 15s
        private const val LOAD_TIMEOUT_MS = 15_000L
        // Retry backoff: 5s → 60s
        private const val RETRY_INITIAL_MS = 5_000L
        private const val RETRY_MAX_MS = 60_000L

        // Cross-manager: shared flag so AppOpenAdManager can avoid showing
        // when interstitial is already showing (and vice versa).
        @Volatile
        var isAnyFullscreenAdShowing = false
    }

    // ==================== Lifecycle / Safety ====================
    @Volatile
    private var isDestroyed = false
    private var sessionId = AtomicLong(0)   // incremented on destroy; callbacks check it

    // ==================== Ad State ====================
    private var interstitialAd: InterstitialAd? = null
    private var loadTime = 0L
    private var lastLoadTime = 0L
    private var lastShowTime = 0L

    // v53: AtomicBoolean CAS guard prevents duplicate InterstitialAd.load() calls
    private val isLoading = AtomicBoolean(false)
    @Volatile
    private var isShowingAd = false

    // v53: pendingShow — set when show() finds stale/no ad; onAdLoaded auto-shows it
    @Volatile
    private var pendingShow = false

    @Volatile
    private var isActivityResumed = false

    // ==================== Retry / Timeout ====================
    private var retryCount = 0
    private val handler = Handler(Looper.getMainLooper())
    private var retryRunnable: Runnable? = null
    private var loadTimeoutRunnable: Runnable? = null

    // ==================== Foreground ====================
    private var lastForegroundAdTime = 0L
    private var wasBackgrounded = false
    private var backgroundTimestamp = 0L

    // ==================== Test Lab ====================
    private var hasAutoShownInterstitial = false
    private var interstitialAutoShowAttempts = 0

    // ==================== Network Callback ====================
    private var networkCallback: ConnectivityManager.NetworkCallback? = null

    private val isTestLab: Boolean by lazy {
        runCatching {
            "true".equals(
                Settings.System.getString(context.contentResolver, "firebase.test.lab"),
                ignoreCase = true
            )
        }.getOrDefault(false)
    }

    private val interstitialAdUnitId by lazy {
        if (isTestLab) TEST_INTERSTITIAL_ID
        else context.getString(com.babelwords.app.R.string.admob_interstitial_id)
    }

    // ==================== Freshness ====================
    private fun isFresh() = interstitialAd != null &&
        (System.currentTimeMillis() - loadTime) < AD_EXPIRY_MS

    private fun shouldRefresh() = interstitialAd != null &&
        (System.currentTimeMillis() - loadTime) >= AD_REFRESH_THRESHOLD_MS

    // ==================== Preload (public) ====================
    fun preloadInterstitial() {
        if (isDestroyed) return
        load(null, isPreload = true)
    }

    // ==================== Show ====================
    fun showInterstitial(activity: Activity) {
        if (isDestroyed) return
        val expectedSession = sessionId.get()

        if (isShowingAd || isAnyFullscreenAdShowing) {
            Log.w(TAG, "showInterstitial: already showing (local=$isShowingAd global=$isAnyFullscreenAdShowing)")
            eventCallback("interstitialFailed", "already_showing")
            return
        }

        // Frequency cap
        val now = System.currentTimeMillis()
        if (now - lastShowTime < 30_000L) {
            Log.w(TAG, "showInterstitial: frequency cap (30s)")
            eventCallback("interstitialFailed", "frequency_capped")
            return
        }

        // Staleness guard: refresh before showing
        if (interstitialAd != null && !isFresh()) {
            Log.w(TAG, "Ad stale — reloading with auto-show")
            eventCallback("interstitialFailed", "ad_stale_reloading")
            pendingShow = true
            load(activity)
            return
        }

        val ad = interstitialAd
        if (ad == null) {
            Log.w(TAG, "No cached ad — triggering load with auto-show")
            eventCallback("interstitialFailed", "no_cached_ad")
            pendingShow = true
            load(activity)
            return
        }

        isShowingAd = true
        isAnyFullscreenAdShowing = true
        lastShowTime = now
        setAudioModeForAd(activity)
        try {
            ad.show(activity)
        } catch (e: Exception) {
            isShowingAd = false
            isAnyFullscreenAdShowing = false
            restoreAudioMode(activity)
            Log.w(TAG, "showInterstitial threw: ${e.message}")
            eventCallback("interstitialFailed", e.message ?: "show_exception")
            // Increment session to invalidate any pending callbacks
            sessionId.incrementAndGet()
            load(null)
        }
    }

    // ==================== 1-step load-and-show ====================
    fun loadInterstitialAndShow(activity: Activity) {
        if (isDestroyed) return
        val cached = interstitialAd
        if (cached != null && isFresh()) {
            Log.d(TAG, "Using cached interstitial")
            // Do NOT null interstitialAd here — let showInterstitial() consume it
            // and the callback null it after successful show.
            showInterstitial(activity)
            load(null)  // refresh after show
            return
        }
        Log.d(TAG, "No cached/fresh interstitial — loading fresh with auto-show")
        pendingShow = true
        load(activity)
    }

    // Backward-compat alias
    fun loadRewardedAndShow(activity: Activity) = loadInterstitialAndShow(activity)

    // ==================== Core Load (single-flight) ====================
    private fun load(activity: Activity?, isPreload: Boolean = false) {
        if (isDestroyed) return
        val expectedSession = sessionId.get()

        // v53: AtomicBoolean CAS guard
        if (!isLoading.compareAndSet(false, true)) {
            Log.d(TAG, "Already loading — skipping duplicate request")
            return
        }

        // Throttle: only advance lastLoadTime when activity is resumed
        // (pre-resume loads don't consume the throttle budget)
        val now = System.currentTimeMillis()
        if (isActivityResumed && now - lastLoadTime < LOAD_THROTTLE_MS) {
            Log.d(TAG, "Load throttled (${(now - lastLoadTime) / 1000}s since last)")
            isLoading.set(false)
            // Schedule a retry after throttle clears
            val delay = LOAD_THROTTLE_MS - (now - lastLoadTime)
            val runnable = Runnable { load(activity, isPreload) }
            retryRunnable = runnable
            handler.postDelayed(runnable, delay)
            return
        }

        if (isActivityResumed) {
            lastLoadTime = now
        }

        Log.d(TAG, "Loading interstitial…")
        val request = getConsentManager()?.buildAdRequest() ?: AdRequest.Builder().build()

        // 15s timeout
        loadTimeoutRunnable?.let { handler.removeCallbacks(it) }
        val timeout = Runnable {
            if (isLoading.get()) {
                Log.w(TAG, "Interstitial load timed out after ${LOAD_TIMEOUT_MS}ms")
                isLoading.set(false)
                interstitialAd = null
                scheduleRetry(RETRY_INITIAL_MS, expectedSession)
            }
        }
        loadTimeoutRunnable = timeout
        handler.postDelayed(timeout, LOAD_TIMEOUT_MS)

        InterstitialAd.load(context, interstitialAdUnitId, request,
            object : InterstitialAdLoadCallback() {
                override fun onAdLoaded(ad: InterstitialAd) {
                    loadTimeoutRunnable?.let { handler.removeCallbacks(it) }
                    if (isDestroyed || sessionId.get() != expectedSession) {
                        ad.fullScreenContentCallback = null
                        isLoading.set(false)
                        return
                    }
                    Log.d(TAG, "✅ Interstitial loaded")
                    interstitialAd = ad
                    loadTime = System.currentTimeMillis()
                    isLoading.set(false)
                    retryCount = 0
                    ad.fullScreenContentCallback = buildCallback(expectedSession)
                    eventCallback("interstitialLoaded", null)

                    // v53: auto-show if pendingShow was set
                    if (pendingShow) {
                        pendingShow = false
                        val act = activity ?: (context as? Activity)
                        if (act != null && !act.isFinishing && !act.isDestroyed) {
                            showInterstitial(act)
                        }
                    } else {
                        maybeAutoShowInterstitial()
                    }
                }
                override fun onAdFailedToLoad(error: com.google.android.gms.ads.LoadAdError) {
                    loadTimeoutRunnable?.let { handler.removeCallbacks(it) }
                    if (isDestroyed || sessionId.get() != expectedSession) {
                        isLoading.set(false)
                        return
                    }
                    Log.w(TAG, "Interstitial load failed: ${error.message} (code=${error.code})")
                    interstitialAd = null
                    isLoading.set(false)
                    eventCallback("interstitialFailed", error.message)

                    val delay = when {
                        error.code == -2 && retryCount == 0 -> 0L
                        error.code == -2 -> (RETRY_INITIAL_MS * (1 shl retryCount)).coerceAtMost(RETRY_MAX_MS)
                        error.code == 3 -> 30_000L
                        else -> RETRY_INITIAL_MS
                    }
                    retryCount++
                    if (retryCount > 5) {
                        retryCount = 0
                        Log.w(TAG, "Retry exhausted after 5 attempts")
                        return
                    }
                    scheduleRetry(delay, expectedSession)
                }
            }
        )
    }

    // ==================== Callbacks ====================
    private fun buildCallback(expectedSession: Long) = object : FullScreenContentCallback() {
        override fun onAdShowedFullScreenContent() {
            if (sessionId.get() != expectedSession) return
            Log.d(TAG, "✅ Interstitial shown")
            isAnyFullscreenAdShowing = true
            if (isTestLab) hasAutoShownInterstitial = true
            interstitialAd = null
            eventCallback("interstitialShown", null)
        }
        override fun onAdDismissedFullScreenContent() {
            if (sessionId.get() != expectedSession) return
            Log.d(TAG, "Interstitial dismissed")
            isShowingAd = false
            isAnyFullscreenAdShowing = false
            restoreAudioMode(context as? Activity)
            eventCallback("interstitialClosed", null)
            load(null)
        }
        override fun onAdFailedToShowFullScreenContent(error: AdError) {
            if (sessionId.get() != expectedSession) return
            Log.w(TAG, "Interstitial show failed: ${error.message}")
            isShowingAd = false
            isAnyFullscreenAdShowing = false
            restoreAudioMode(context as? Activity)
            interstitialAd = null
            eventCallback("interstitialFailed", error.message)
            load(null)
        }
    }

    // ==================== Retry ====================
    private fun scheduleRetry(delayMs: Long, expectedSession: Long) {
        retryRunnable?.let { handler.removeCallbacks(it) }
        val runnable = Runnable {
            if (!isDestroyed && sessionId.get() == expectedSession) {
                load(null)
            }
        }
        retryRunnable = runnable
        handler.postDelayed(runnable, delayMs)
        Log.d(TAG, "Retrying interstitial load in ${delayMs}ms (attempt $retryCount)")
    }

    fun cancelRetries() {
        retryRunnable?.let { handler.removeCallbacks(it) }
        retryRunnable = null
        loadTimeoutRunnable?.let { handler.removeCallbacks(it) }
        loadTimeoutRunnable = null
    }

    // ==================== Foreground Ad ====================
    fun onActivityResumed(activity: Activity) {
        if (isDestroyed) return
        isActivityResumed = true
        val now = System.currentTimeMillis()

        // Execute any deferred pendingShow
        if (pendingShow) {
            pendingShow = false
            showInterstitial(activity)
            return
        }

        if (wasBackgrounded) {
            val inBackground = now - backgroundTimestamp
            wasBackgrounded = false
            if (inBackground >= 3_000L && now - lastForegroundAdTime >= FOREGROUND_AD_COOLDOWN_MS) {
                lastForegroundAdTime = now
                Log.d(TAG, "Showing foreground interstitial")
                loadInterstitialAndShow(activity)
                return
            }
        }
    }

    fun onActivityPaused() {
        isActivityResumed = false
        wasBackgrounded = true
        backgroundTimestamp = System.currentTimeMillis()
    }

    // ==================== Network Callback ====================
    fun registerNetworkCallback() {
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            ?: return
        val request = android.net.NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        val callback = object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: android.net.Network) {
                if (isDestroyed) return
                // Auto-reload if we have a stale ad or no ad at all.
                // CRITICAL: NetworkCallback runs on ConnectivityThread (background).
                // InterstitialAd.load() requires main thread — must dispatch via handler.
                if (interstitialAd == null || !isFresh()) {
                    Log.d(TAG, "Network available — auto-reloading interstitial (dispatch to main thread)")
                    handler.post { load(null) }
                }
            }
        }
        networkCallback = callback
        cm.registerNetworkCallback(request, callback)
        Log.d(TAG, "Network callback registered")
    }

    fun unregisterNetworkCallback() {
        networkCallback?.let { callback ->
            val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            cm?.unregisterNetworkCallback(callback)
            networkCallback = null
        }
    }

    // ==================== Test Lab Auto-show ====================
    private fun maybeAutoShowInterstitial() {
        if (!isTestLab) return
        // Fail-safe: gate auto-show on successful test-device registration.
        // If registration failed, suppress auto-show to avoid real-creative impressions
        // (invalid traffic). Prefer no impression over a risky one.
        if (!BabelWordsApplication.isTestDeviceRegistrationActive) {
            Log.w(TAG, "🧪 Test Lab: auto-show suppressed — test-device registration not active")
            return
        }
        if (hasAutoShownInterstitial) return
        if (interstitialAutoShowAttempts >= MAX_AUTO_SHOW_ATTEMPTS) {
            Log.w(TAG, "🧪 Test Lab: interstitial auto-show gave up")
            return
        }
        val activity = context as? Activity ?: return
        if (activity.isFinishing || activity.isDestroyed) return
        interstitialAutoShowAttempts++
        Log.i(TAG, "🯪 Test Lab: auto-showing interstitial (attempt $interstitialAutoShowAttempts)")
        activity.runOnUiThread { showInterstitial(activity) }
    }

    // ==================== Audio ====================
    private fun setAudioModeForAd(activity: Activity?) {
        val am = activity?.getSystemService(Context.AUDIO_SERVICE) as? AudioManager ?: return
        am.mode = AudioManager.MODE_NORMAL
    }

    private fun restoreAudioMode(activity: Activity?) {
        val am = activity?.getSystemService(Context.AUDIO_SERVICE) as? AudioManager ?: return
        am.mode = AudioManager.MODE_NORMAL
    }

    // ==================== Destroy ====================
    fun destroy() {
        isDestroyed = true
        sessionId.incrementAndGet()   // invalidate all pending callbacks
        cancelRetries()
        unregisterNetworkCallback()
        interstitialAd?.fullScreenContentCallback = null
        interstitialAd = null
        isLoading.set(false)
        pendingShow = false
        isAnyFullscreenAdShowing = false
        isShowingAd = false
        Log.d(TAG, "Destroyed")
    }

    // ==================== Queries ====================
    fun isInterstitialReady() = !isDestroyed && interstitialAd != null && isFresh()
    fun isRewardedReady() = isInterstitialReady()
    fun isInitialized() = !isDestroyed
}

File: ios/BabelWords/Managers/AdMobManager.swift
import Foundation
@preconcurrency import GoogleMobileAds
import Network
import UIKit
import AVFoundation

// MARK: - Ad loading

protocol InterstitialAdLoading {
    func load(
        withAdUnitID adUnitID: String,
        request: GADRequest,
        completionHandler: @escaping @Sendable (GADInterstitialAd?, Error?) -> Void
    )
}

struct GADInterstitialAdLoader: InterstitialAdLoading {
    func load(
        withAdUnitID adUnitID: String,
        request: GADRequest,
        completionHandler: @escaping @Sendable (GADInterstitialAd?, Error?) -> Void
    ) {
        GADInterstitialAd.load(
            withAdUnitID: adUnitID,
            request: request,
            completionHandler: completionHandler
        )
    }
}

// MARK: - Shared fullscreen state

final class FullscreenAdState: @unchecked Sendable {
    private let lock = NSLock()
    private var showing = false

    var isShowing: Bool {
        lock.lock()
        defer { lock.unlock() }
        return showing
    }

    func setShowing(_ showing: Bool) {
        lock.lock()
        defer { lock.unlock() }
        self.showing = showing
    }
}

/// Unified interstitial ad manager for iOS. Replaces the Android `AdMobManager`.
final class AdMobManager: NSObject, @unchecked Sendable {
    private let TAG = "AdMobManager"

    private static let testInterstitialID = "ca-app-pub-3940256099942544/4411468910"
    private static let maxAutoShowAttempts = 3
    private static let foregroundAdCooldown: TimeInterval = 5 * 60
    private static let adExpiry: TimeInterval = 45 * 60
    private static let adRefreshThreshold: TimeInterval = 40 * 60
    private static let loadThrottle: TimeInterval = 30
    private static let loadTimeout: TimeInterval = 15
    private static let retryInitial: TimeInterval = 5
    private static let retryMax: TimeInterval = 60
    private static let retryMaxCount = 10

    /// Cross-manager state so App Open ads cannot collide with interstitials.
    static let fullscreenAdState = FullscreenAdState()

    var eventCallback: ((String, String?) -> Void)?
    private var getConsentManager: () -> ConsentManager?
    private var adLoader: InterstitialAdLoading

    /// Called when the SDK's completion handler arrives on a background thread.
    /// Debug builds assert for early detection; Release builds recover safely.
    var offMainThreadHandler: (String) -> Void = {
        #if DEBUG
        assertionFailure($0)
        #else
        print("[AdMobManager] \($0)")
        #endif
    }

    private var interstitial: GADInterstitialAd?
    private var loadTime: Date?
    private var lastLoadTime: Date?
    private var lastShowTime: Date? {
        get { UserDefaults.standard.object(forKey: "interstitial_last_show_time") as? Date }
        set { UserDefaults.standard.set(newValue, forKey: "interstitial_last_show_time") }
    }

    private var isLoading = false
    private var isShowingAd = false
    private var pendingShow = false
    private var isActivityResumed = false

    private var retryCount = 0
    private var retryTask: Task<Void, Never>?
    private var loadTimeoutTask: Task<Void, Never>?

    private var lastForegroundAdTime: Date = .distantPast
    private var wasBackgrounded = false
    private var backgroundTimestamp: Date?

    private var hasAutoShownInterstitial = false
    private var interstitialAutoShowAttempts = 0

    private var networkMonitor: NWPathMonitor?

    private var isTestLab: Bool {
        ProcessInfo.processInfo.environment["FIREBASE_TEST_LAB"] == "true"
    }

    private var interstitialAdUnitID: String {
        isTestLab ? AdMobManager.testInterstitialID : "ca-app-pub-9991891515643313/7320741331"
    }

    init(
        getConsentManager: @escaping () -> ConsentManager? = { nil },
        adLoader: InterstitialAdLoading = GADInterstitialAdLoader()
    ) {
        self.getConsentManager = getConsentManager
        self.adLoader = adLoader
        super.init()
    }

    // MARK: - Freshness

    private func isFresh() -> Bool {
        guard interstitial != nil, let loadTime = loadTime else { return false }
        return Date().timeIntervalSince(loadTime) < AdMobManager.adExpiry
    }

    private func shouldRefresh() -> Bool {
        guard let loadTime = loadTime else { return false }
        return Date().timeIntervalSince(loadTime) >= AdMobManager.adRefreshThreshold
    }

    // MARK: - Preload / Show / Load-and-show

    func preloadInterstitial() {
        load(presentingViewController: nil, isPreload: true)
    }

    func showInterstitial(from viewController: UIViewController) {
        if isShowingAd || AdMobManager.fullscreenAdState.isShowing {
            print("[\(TAG)] showInterstitial: already showing")
            eventCallback?("interstitialFailed", "already_showing")
            return
        }

        if let lastShow = lastShowTime, Date().timeIntervalSince(lastShow) < 30 {
            print("[\(TAG)] showInterstitial: frequency cap")
            eventCallback?("interstitialFailed", "frequency_capped")
            return
        }

        if interstitial != nil, !isFresh() {
            print("[\(TAG)] Ad stale — reloading with auto-show")
            eventCallback?("interstitialFailed", "ad_stale_reloading")
            pendingShow = true
            load(presentingViewController: viewController)
            return
        }

        guard let ad = interstitial else {
            print("[\(TAG)] No cached ad — loading with auto-show")
            eventCallback?("interstitialFailed", "no_cached_ad")
            pendingShow = true
            load(presentingViewController: viewController)
            return
        }

        isShowingAd = true
        AdMobManager.fullscreenAdState.setShowing(true)
        lastShowTime = Date()
        setAudioModeForAd()

        ad.fullScreenContentDelegate = self
        ad.present(fromRootViewController: viewController)
    }

    func loadInterstitialAndShow(from viewController: UIViewController) {
        if interstitial != nil, isFresh() {
            showInterstitial(from: viewController)
            load(presentingViewController: nil, isPreload: true)
            return
        }
        pendingShow = true
        load(presentingViewController: viewController)
    }

    func loadRewardedAndShow(from viewController: UIViewController) {
        // Backward-compat alias: rewarded video removed, now uses interstitial.
        loadInterstitialAndShow(from: viewController)
    }

    // MARK: - Core load

    private func load(presentingViewController: UIViewController?, isPreload: Bool = false) {
        guard !isLoading else {
            print("[\(TAG)] Already loading — skipping duplicate request")
            return
        }

        let now = Date()
        if isActivityResumed, let last = lastLoadTime, now.timeIntervalSince(last) < AdMobManager.loadThrottle {
            print("[\(TAG)] Load throttled")
            isLoading = false
            let delay = AdMobManager.loadThrottle - now.timeIntervalSince(last)
            retryTask = Task { [weak self] in
                do {
                    try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                } catch {
                    return
                }
                DispatchQueue.main.async { [weak self] in
                    guard let self = self, !self.isLoading else { return }
                    self.load(presentingViewController: presentingViewController, isPreload: isPreload)
                }
            }
            return
        }

        if isActivityResumed {
            lastLoadTime = now
        }

        if isPreload && isSlowOrMeteredNetwork() {
            print("[\(TAG)] Skipping preload — slow or metered network")
            isLoading = false
            return
        }

        print("[\(TAG)] Loading interstitial…")
        isLoading = true

        let request = getConsentManager()?.buildAdRequest() ?? GADRequest()

        loadTimeoutTask?.cancel()
        loadTimeoutTask = Task { [weak self] in
            do {
                try await Task.sleep(nanoseconds: UInt64(AdMobManager.loadTimeout * 1_000_000_000))
            } catch {
                return
            }
            DispatchQueue.main.async { [weak self] in
                guard let self = self, self.isLoading else { return }
                print("[\(self.TAG)] Interstitial load timed out")
                self.isLoading = false
                self.interstitial = nil
                self.scheduleRetry(delay: AdMobManager.retryInitial)
            }
        }

        adLoader.load(
            withAdUnitID: interstitialAdUnitID,
            request: request
        ) { [weak self] ad, error in
            let errorDescription = error?.localizedDescription
            let errorCode = (error as NSError?)?.code
            guard Thread.isMainThread else {
                self?.offMainThreadHandler(
                    "[\(self?.TAG ?? "AdMobManager")] GADInterstitialAd completion called off main thread"
                )
                Task { @MainActor [weak self] in
                    guard let self = self else { return }
                    self.handleOffMainInterstitialCallback(
                        errorDescription: errorDescription,
                        errorCode: errorCode
                    )
                }
                return
            }
            self?.handleInterstitialLoad(
                ad: ad,
                errorDescription: errorDescription,
                errorCode: errorCode
            )
        }
    }

    private func handleOffMainInterstitialCallback(errorDescription: String?, errorCode: Int?) {
        print("[\(TAG)] Ignoring off-main interstitial callback\(errorDescription.map { ": \($0)" } ?? "")")
        loadTimeoutTask?.cancel()
        isLoading = false
        interstitial = nil
        scheduleRetry(delay: retryDelay(for: errorCode))
    }

    private func handleInterstitialLoad(
        ad: GADInterstitialAd?,
        errorDescription: String?,
        errorCode: Int?
    ) {
        loadTimeoutTask?.cancel()
        isLoading = false

        if let errorDescription = errorDescription {
            print("[\(TAG)] Interstitial load failed: \(errorDescription)")
            interstitial = nil
            eventCallback?("interstitialFailed", errorDescription)
            scheduleRetry(delay: retryDelay(for: errorCode))
        } else if let ad = ad {
            print("[\(TAG)] Interstitial loaded")
            interstitial = ad
            loadTime = Date()
            retryCount = 0
            ad.fullScreenContentDelegate = self
            eventCallback?("interstitialLoaded", nil)

            if pendingShow {
                pendingShow = false
                if let vc = topViewController() {
                    showInterstitial(from: vc)
                }
            } else {
                maybeAutoShowInterstitial()
            }
        }
    }

    private func retryDelay(for code: Int?) -> TimeInterval {
        switch code {
        case -2 where retryCount == 0:
            return 0
        case -2:
            return min(AdMobManager.retryInitial * pow(2.0, Double(retryCount)), AdMobManager.retryMax)
        case 3:
            return 30
        default:
            return AdMobManager.retryInitial
        }
    }

    // MARK: - Retry

    private func scheduleRetry(delay: TimeInterval) {
        retryCount += 1
        if retryCount > AdMobManager.retryMaxCount {
            print("[\(TAG)] Retry limit reached")
            eventCallback?("interstitialFailed", "retry_limit_reached")
            return
        }
        retryTask?.cancel()
        retryTask = Task { [weak self] in
            do {
                try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            } catch {
                return
            }
            DispatchQueue.main.async { [weak self] in
                guard let self = self, !self.isLoading else { return }
                self.load(presentingViewController: nil, isPreload: true)
            }
        }
        print("[\(TAG)] Retrying interstitial load in \(delay)s (attempt \(retryCount)/\(AdMobManager.retryMaxCount))")
    }

    private func cancelRetries() {
        retryTask?.cancel()
        retryTask = nil
        loadTimeoutTask?.cancel()
        loadTimeoutTask = nil
    }

    // MARK: - Foreground ad

    func onActivityResumed(_ viewController: UIViewController) {
        isActivityResumed = true

        if pendingShow {
            pendingShow = false
            showInterstitial(from: viewController)
            return
        }

        if wasBackgrounded, let backgroundTimestamp = backgroundTimestamp {
            let inBackground = Date().timeIntervalSince(backgroundTimestamp)
            wasBackgrounded = false
            if inBackground >= 3 && Date().timeIntervalSince(lastForegroundAdTime) >= AdMobManager.foregroundAdCooldown {
                lastForegroundAdTime = Date()
                print("[\(TAG)] Showing foreground interstitial")
                loadInterstitialAndShow(from: viewController)
            }
        }
    }

    func onActivityPaused() {
        isActivityResumed = false
        wasBackgrounded = true
        backgroundTimestamp = Date()
    }

    // MARK: - Network monitoring

    func registerNetworkCallback() {
        let monitor = NWPathMonitor()
        monitor.pathUpdateHandler = { [weak self] path in
            Task { @MainActor [weak self] in
                guard let self = self else { return }
                if path.status == .satisfied,
                   self.interstitial == nil || !self.isFresh() {
                    print("[\(self.TAG)] Network available — auto-reloading")
                    self.load(presentingViewController: nil, isPreload: true)
                }
            }
        }
        monitor.start(queue: DispatchQueue.global(qos: .utility))
        networkMonitor = monitor
    }

    func unregisterNetworkCallback() {
        networkMonitor?.cancel()
        networkMonitor = nil
    }

    private func isSlowOrMeteredNetwork() -> Bool {
        guard let monitor = networkMonitor else { return false }
        let path = monitor.currentPath
        let isWifi = path.usesInterfaceType(.wifi)
        let isExpensive = path.isExpensive
        return path.status == .satisfied && !isWifi && isExpensive
    }

    // MARK: - Test Lab auto-show

    private func maybeAutoShowInterstitial() {
        guard isTestLab else { return }
        guard AppDelegate.isTestDeviceRegistrationActive else {
            print("[\(TAG)] Test Lab: auto-show suppressed — test-device registration not active")
            return
        }
        guard !hasAutoShownInterstitial else { return }
        guard interstitialAutoShowAttempts < AdMobManager.maxAutoShowAttempts else { return }

        interstitialAutoShowAttempts += 1
        print("[\(TAG)] Test Lab: auto-showing interstitial (attempt \(interstitialAutoShowAttempts))")
        // We need the topmost view controller; use a helper to find it.
        if let vc = topViewController() {
            showInterstitial(from: vc)
        }
    }

    private func topViewController() -> UIViewController? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let root = windowScene.windows.first?.rootViewController else { return nil }
        var vc = root
        while let presented = vc.presentedViewController { vc = presented }
        return vc
    }

    // MARK: - Audio

    private func setAudioModeForAd() {
        try? AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
        try? AVAudioSession.sharedInstance().setActive(true)
    }

    private func restoreAudioMode() {
        try? AVAudioSession.sharedInstance().setCategory(.playAndRecord, mode: .voiceChat, options: [.defaultToSpeaker])
        try? AVAudioSession.sharedInstance().setActive(true)
    }

    // MARK: - Destroy

    func destroy() {
        cancelRetries()
        unregisterNetworkCallback()
        interstitial?.fullScreenContentDelegate = nil
        interstitial = nil
        isLoading = false
        pendingShow = false
        AdMobManager.fullscreenAdState.setShowing(false)
        isShowingAd = false
    }

    // MARK: - Queries

    func isInterstitialReady() -> Bool { interstitial != nil && isFresh() }
    func isRewardedReady() -> Bool { isInterstitialReady() }
    func isInitialized() -> Bool { true }
}

// MARK: - GADFullScreenContentDelegate

extension AdMobManager: GADFullScreenContentDelegate {
    func adDidRecordImpression(_ ad: GADFullScreenPresentingAd) {
        guardMainThread { [weak self] in
            self?.adDidRecordImpressionOnMain()
        }
    }

    private func adDidRecordImpressionOnMain() {
        AdMobManager.fullscreenAdState.setShowing(true)
    }

    func adDidRecordClick(_ ad: GADFullScreenPresentingAd) {
        let adUnitID = interstitialAdUnitID
        guardMainThread {
            AnalyticsManager.logAdClicked(adUnit: adUnitID, adFormat: "interstitial")
        }
    }

    func adWillPresentFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        guardMainThread { [weak self] in
            self?.adWillPresentFullScreenContentOnMain()
        }
    }

    private func adWillPresentFullScreenContentOnMain() {
        print("[\(TAG)] Interstitial shown")
        AdMobManager.fullscreenAdState.setShowing(true)
        hasAutoShownInterstitial = true
        interstitial = nil
        eventCallback?("interstitialShown", nil)
        AnalyticsManager.logAdImpression(adUnit: interstitialAdUnitID, adFormat: "interstitial")
    }

    func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        guardMainThread { [weak self] in
            self?.adDidDismissFullScreenContentOnMain()
        }
    }

    private func adDidDismissFullScreenContentOnMain() {
        print("[\(TAG)] Interstitial dismissed")
        isShowingAd = false
        AdMobManager.fullscreenAdState.setShowing(false)
        restoreAudioMode()
        eventCallback?("interstitialClosed", nil)
        load(presentingViewController: nil, isPreload: true)
    }

    func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        let description = error.localizedDescription
        let adUnitID = interstitialAdUnitID
        guardMainThread { [weak self] in
            self?.adDidFailToPresentOnMain(description: description, adUnitID: adUnitID)
        }
    }

    private func adDidFailToPresentOnMain(description: String, adUnitID: String) {
        print("[\(TAG)] Interstitial show failed: \(description)")
        isShowingAd = false
        AdMobManager.fullscreenAdState.setShowing(false)
        restoreAudioMode()
        interstitial = nil
        eventCallback?("interstitialFailed", description)
        AnalyticsManager.logAdFailed(adUnit: adUnitID, error: description)
        load(presentingViewController: nil, isPreload: true)
    }

    private func guardMainThread(_ operation: @escaping @Sendable () -> Void) {
        if Thread.isMainThread {
            operation()
        } else {
            DispatchQueue.main.async(execute: operation)
        }
    }
}
